### PR TITLE
GUI reconfiguration one size fits all [don't merge]

### PR DESCRIPTION
**Pull request by ErichZimmer**
Dear Mr. Vennemann,

As requested, I moved the conversation here. To start, here is the finalized configuration. Note, I kept the user function in the same frame due to issues, so any scripts/code should be imported and not typed due to low space. I added a a flip array option for the u and v components so the user now has nearly total control of how the u/v components are saved. Keyboard shortcuts are going to be added for added efficiency. So far, there are minor bugs/aesthetic issues, but those should be fixed in some 5 days. 

**ensemble correlation**
I looked at the theory behind ensemble correlation and it looks pretty simple for single pass evaluations. How ever, multi-pass evaluations might be pretty challenging. I read through nearly all of the PIVlab and OpenPIV google forums and learned a lot from them, so maybe, just maybe, I can get something to work for once :P

GUI_General

GUI_Preproc

As mentioned above, the user has nearly total control of the u/v components. This allows the user to correctly place the vectors with the origin in the top left (even though processing occurs bottom left?).
GUI_PIV

GUI_Preproc2

GUI_Post

GUI_Plot

Somehow, the frames, spacers, and buttons need to be omitted from the laboratory log book.
GUI_Log

Regards,
Erich Zimmer

PS, sorry if this isn't too informative/concise as it is currently 4 hours past my bed time (stayed up half the night so far :)). Also, don't merge as I won't be pushing the new code for some 5 days after testing and checking for mistakes.

### CODE REVIEW

**Comment by eguvep** [comment]
**user function**
I think, this if fine. The frame is still good enough for writing three lines of code. For more, one should anyway use a proper editor. Let's do it as Kent Thompson said: »Make each program do one thing well. To do a new job, build a fresh rather than complicate old programs by adding new "features".«

**ensemble correlation**
I once implemented it in JPIV: https://github.com/eguvep/jpiv/blob/master/src/jpiv2/SinglePixelCorr.java But I repeat myself: Until February there is no time for rewriting this in Python... unfortunately

The proposed changes look very good to me. Take your time, and get enough sleep ;-)

When you are ready for merging, please let me know. Please pull upstream, solve conflicts locally and then push into your branch, just before notifying me. This should minimize merge conflicts, possibly even enabling an automatic merge.

Take care!

Peter

**Comment by alexlib** [comment]
> **ensemble correlation**
> I once implemented it in JPIV: https://github.com/eguvep/jpiv/blob/master/src/jpiv2/SinglePixelCorr.java But I repeat myself: Until February there is no time for rewriting this in Python... unfortunately

I think this should be discussed at openpiv-python forum, right? I'd like to see ensemble correlation there, maybe also with other enhanced features

**Comment by eguvep** [comment]
Yes, Alex, definitely. We should clearly separate working code from the user interface, and as far as possible also the discussion about it.

**Comment by alexlib** [comment]
In order to process this pull request effectively, @ErichZimmer please first check locally on your machine that you can merge the master from OpenPIV - your master branch is `This branch is 2 commits ahead, 47 commits behind OpenPIV:master.`

**Comment by ErichZimmer** [comment]
Dear Prof. Liberzon,

Does it make a difference if I pull upstream first, copy and paste the differences with the new modifications, push modified branch, and then check for merge conflicts? Does it matter if I pull directly from the OpenPIV/openpiv_tk_gui or is there a more effective/better solution? So far, I am simply following the contribution instructions in the readme.m file plus some additional stuff from GitHub instructions. I'm still new to GitHub and I am still learning about merge conflicts so this gets a little confusing at times.

**Comment by alexlib** [comment]
> Dear Prof. Liberzon,
> 
> Does it make a difference if I pull upstream first, copy and paste the differences with the new modifications, push modified branch, and then check for merge conflicts? Does it matter if I pull directly from the OpenPIV/openpiv_tk_gui or is there a more effective/better solution? So far, I am simply following the contribution instructions in the readme.m file plus some additional stuff from GitHub instructions. I'm still new to GitHub and I am still learning about merge conflicts so this gets a little confusing at times.


Basically these are the instructions for the case without write permissions from the point 6 and on:

```
See, if there are updates in the upstream repository and save them in your local branch upstream/master:
git fetch upstream
Merge possible upstream changes into your local master branch:
git merge upstream/master
If there are merge conflicts, use git status and git diff for displaying them. Git marks conflicts in your files, as described in the Github documentation on solving merge conflicts. After resolving merge conflicts, upload everything:
git add .
git commit -m 'A meaningful comment.'
git push
Propose your changes to the upstream developer by creating a pull-request, as described in the Github documentation for creating a pull-request from a fork. (Basically just pressing the »New pull request« button.)
```

That's what you need to do now so we could continue with the pull request: 
1. Pull from `upstream` remote repository, which is for you the `OpenPIV/openpiv_tk_gui` and merge upstream changes into your master (or another branch that you want to suggest as a pull request)
2. Verify that tests pass (if there are tests) and that your changes do not break somebody else changes (that occurred while you were working on a new feature)
3. If all merged fine - just push into your repository `master` branch that you suggested as a pull request. The `upstream` will see in the pull request only the differences between itself and your changes on top of the present upstream master. 

Hope it's clear. You do very good job and you follow the instructions very well, thanks for your contributions.

**Comment by ErichZimmer** [comment]
Dear Prof. Liberzon,

Do you remember which feature I broke with one of my additions? I am creating a test to verify that everything is working and comparing scripts that were written before my contributions to make sure I didn't mess anything up.

**Comment by alexlib** [comment]
I don't know - there are simply merge conflicts that only the author knows how to fix. Some lines of code are different now from the state you started your changes. Please review carefully and remove the merge conflicts, merge locally, try the usual testing procedure you use and then push to your github repo.

**Comment by eguvep** [comment]
Git will indicate merge conflicts by inserting comments into your source files. Just scroll through the files. Most editors even highlight the git comments green/red. Sometimes, it is just an additional blank line, which is easy to fix. Sometimes there are two simultaneously changed, competing source blocks. So the author must try to understand both changes, decide, which one to keep or even merge them to a new code block, that is satisfying both changes.
Git will recognize, that you removed all git-comments and will not complain anymore. 
Success!
Peter

**Comment by eguvep** [comment]
... sometimes there are files causing merge conflicts, that should not even be in the repository. You can simply remove them. `git rm filenames`. These are frequently automatically generated files, when executing some code directly in the repository directory or some backup files from the IDE. Strictly, this files should be blacklisted. For whatever reason, there are jupyter leftovers in our repositoy.

**Comment by ErichZimmer** [comment]
Dear Mr. Vennemann,

**new way of handling PIV results**
I was playing around with saving/loading sessions and I ended up with an added dictionary that contains raw and processed results. This allows the whole session (excluding image pairs) to be saved and exported. However, I am not planning to implement it unless if you want it because I think it stretches past our ideals a little bit. For instance, the listbox widget on the right of the GUI will show the dictionary key of the processed image pairs (when selecting image pairs, it sill shows the file paths of the images). Additionally, it adds slightly more code to the  GUI, multiprocessing, and postprocessing with no real enhancement in efficiency. To append the results from the MultiProcessing class, the results can either be saved, appended, or both. A slight problem could be encountered with very large amount of image pairs where the dictionary could slow down. So far, I tested it with 1300 image pairs and it performed slighly faster than loading a .txt or .vec file. 
The dictionary is set up like this:
let 'i' equal the index key number for a data set.
[i][0] = x array
[i][1] = y array
[i][2] = raw u array
[i][3] = raw v array
[i][4] = signal to noise ratio
[i][5] = typevector (if values were marked NaN or interpolated)
[i][6] = mask (for areas to not be analyzed/interpolated)
[i][7] = postprocessed u array
[i][8] = postprocessed v array
The arrays are appended in the MultiProcessing class at the end of each iteration. Each index key could be saved as a .vec or .txt file in standard 'x, y, u, v, typevector (mask)' format. In postprocessing, the u and v results are appended as well as the typevector (mask). 

**GUI dropdown menus**
To allow the GUI to be smaller, I added dropdown menus. Instead of having 14 riders/buttons, it is now reduced to 10 buttons. I like the appearance of the dropdown menus, but now the user has to click everything twice. Here is an example of the dropdown menu bar.

dropdown_menu_bar

I might get rid of the dropdown menus and instead replace them with buttons and nested notebooks. For example, instead of having three preprocessing dropdown menus, it would be a preprocessing button that selects a nested notebook that would have a 'background', 'masks' and 'filters' rider. This would, in my opinion, increase the efficiency compared to the dropdown menus as the user wouldn't have to click twice to get to every menu. (hope this is making sense)

Regards,
Erich Zimmer

**Comment by eguvep** [comment]
Dear Erich,

**GUI drop down menus**
I think, the drop down menus are a straightforward way of grouping riders that are somehow related to each other. I would not think, that nested riders are more efficient, because the two clicks on the drop down menu are very close together. I think, this is a good solution for integrating a larger amount of parameters in a way, that users intuitively understand.

**new way of handling PIV results**
Your suggestion is definitely a clearer way of saving the results. The standard 'x, y, u, v, multipurpose' format is kind of the »smallest common denominator« of all PIV software. That is, why I adopted this format for the first versions. It is also important to support this format by providing im- and export functions. 

But a future format decision, is also a decision we can not revert or change easily. As soon as users saved their results in this format, it would not be fair to change it again. We also do not want to care about backwards compatibility for yet another format. This would slow down testing and would clutter the code. 

So I suggest to discuss this very carefully. Probably the format question is rather a question for OpenPIV rather than openpiv_tk_gui. What do you think, @alexlib? Do you remember, if this topic already has been discussed elsewhere? It would be efficient, if everyone in the OpenPIV-Python community would use the same format, instead of converting it from and into own formats for every tool or lib. This will only cause errors and useless overhead. Sticking to the current format, however, means to live with its shortcomings.

- The standard 'x, y, u, v, multipurpose' format is compatible to virtually all OpenPIV terminal script, Jupyter notebook, and other PIV software. Is it a problem to break this compatibility? Is it good enough to provide an import/export function (like others do)?
- If we want to change something, what is the best format?
   + Keeping the standard format? (Not very versatile, neither efficient, but human readable, extremely compatible to virtually every plotting program like Tecplot, or terminal program like awk or gnuplot or MATLAB or even Excel (Whoever wants to import PIV data to Excel?). 
   + A dictionary? (Part of the standard library, well known by all programmers, more versatile than an array,  serializable in very common JSON format, ...)
   + A pandas data frame? (More efficient, direct and simple usage of all pandas functionality, ...). We actually already use dataframes in the GUI ...
   + NetCDF format? (Extremely efficient, developed for research, dependent on a very heavy library, not so straightforward to use, ...)
   + Anything else?

To sum up: I completely understand your wish for another, more versatile format and welcome it. I suggest to start a discussion about a possible format one level higher than the GUI. But we may start the discussion with a suggestion or options. My favourites are a dictionary (as suggested by Erich) and a Pandas DataFrame (because we anyway use that already).

What do you think?

**Comment by alexlib** [comment]
Suggest to ask on the openpiv-users forum. My position is that we are free to decide for every sub-project whatever format is convenient as long as we provide reading/writing/translation to other formats. 
At the moment the only post-processing tools for OpenPIV I know are: Spatial toolbox in Matlab that reads MAT files, VEC files (like Tecplot or Insight ASCII files with a header and x,y,u,v,quality) or TXT files (x,y,u,v) without headers, PIVMAT - we provide a translation from TXT or VEC to Matlab and what we write now is PIVPy - in Python using Xarray format. It's basically Pandas DataFrame but much better for multi-dimensional format with metadata. I recommend you to take a look at https://github.com/alexlib/pivpy and xarray. It is well developed now and it can be saved as HDF5, NetCDF, CSV files, etc. In this sense these will work as smooth as Pandas DataFrames, but with additional features like having attributes in the DataFrame itself. 

I vote for xarray data format and storage in CSV (through Pandas) or NetCDF or HDF5 (whatever is more convenient to install). We will need also to provide Matlab writer/reader for the translation to Spatial Toolbox and PIVMAT

**Comment by eguvep** [comment]
I wasn't aware of xarray – seems to be a good choice!

**Comment by ErichZimmer** [comment]
Xarray looks really nice when I was going through the PivPy scripts. If we store PIV results in an xarray data frame or dictionary, this would allow the user to post-process the PIV results and clear/re-post-process them again to refine the settings. I'll see what I can do with my large amount of free time. ;)

**Comment by eguvep** [comment]
&#128077;

**Comment by ErichZimmer** [comment]
Dear Prof. Vennemann,

I refined the GUI a little bit today, but I am not sure if it's in the right direction. First off, I added popup warnings to inform the user of some user error inputs. I also added a delimiter to the saving preferences. To make the GUI a little less command-line dependent, I added a file filter hint at the top right of the GUI.

General

Additionally, I changed up the PIV settings to make it look a little more mature, but I am not sure if that is going in the right direction too. So before I devote more time on it, here is what I put together.

algorithms

Capture

validation

To start the evaluation, the user can simply click the »start processing« menu. This would bring up a hidden tab which has some information on it (i.e. total amount of frames, progress bar, »stop processing« button). More importantly, it disables all PIV riders to stop the user from changing the settings while the evaluation is in progress. This created a severe conflict on my windows 10 laptop.

In conclusion, I think it would make the gui look nicer to spread out all those widgets between several menus. With the addition of the custom windowing option, an advanced user would have more control over the windowing between passes at the expense of it being a little finicky. Finally, if the user wants to spread the widgets to remove most of the unused space, a checkbox could be created but has an additional expensive of more complicated code. As a great hike I met once said, 'everything comes with a compromise.' :)

Regards, 
Erich Zimmer

**Comment by eguvep** [comment]
Dear Erich,
**pop up warnings**
I think, this is good for the most common errors, because the work flow for beginners will be faster with the warnings. But we should not catch every possible error, because every warning makes the code longer and less clear.
**filter hint**
Very good idea! Especially, because the filter concept is very uncommon – although very versatile and efficient.
**processing menu**
This indeed makes a very professional impression to me. Thanks a lot in the name of the OpenPIV community for devoting so much creativity into the GUI!
**spreading widgets**
I am afraid, I didn't get the idea. Could you explain that a little bit more?

**Comment by ErichZimmer** [comment]
Well, I'm approaching what a pre-college novice python programmer can do, so I'll simply just tighty things up and keep it at that. I really want to try adding a dictionary/xarray data frame, but I'm having plenty of troubles with that. Additionally, I couldn't get a few of the preprocessing features (e.g. object masking, intensity normalization, contrast stretching) and the stop processing button to work properly, so those would be up to the community to contribute (if they deem those features necessary). Because of those issues and the large amounts of debugging, I decided to have a single master preprocessing frame instead of three, as I removed the features plagued with multiple bugs. Now, I just have to make everything look 'nicer' and do quite a bit of testing and I will send a pull request your way. ;)

**Comment by ErichZimmer** [comment]
FWI, this is the general direction I am going with the preprocessing filters. (please ignore canvas background change)

preproc

**Comment by ErichZimmer** [comment]
Dear Prof. Vennemann,

Do we still want to put all images and processing data in a dictionary and/or xarray data frame? I am about finished with the new additions that actually function properly and are still being tested (me being paranoid from the last merge). I personally would like to try it out, but I'm actually running out of time for once as my large amount of 'free time' is  slowly getting smaller.

I reread all of the posts so far to make sure I won't mess anything up and help clean up some of the Jupyter notebook leakage that made it into the repository. I am also comparing my contributions to the code written before I came along for a beginner programmer's best/worst nightmare to make sure that I get all my stuff as clean and good as possible. One problem that could pop up is with the auto documentation of openpivparams. Somehow, it needs to know to not include the frames and spacers but include 'sub parameters' that are placed in the nested LabelFrames and Panedwindows.

**Comment by alexlib** [comment]
Xarray is better for the data, not so much for the images. 
If you share with me your branch @ErichZimmer where you need some help with Xarray - I'd gladly help. I work now on pivpy a bit and have gathered some ideas how to use it more effectively. I believe that pivpy already has most of the needed functionality

**Comment by eguvep** [comment]
Dear Erich,

I think the GUI is pretty usable, already! With a whole bunch of very useful contributions by you! So there is no need for hurrying. Focus on features you really need yourself for your research, and only as long as your time allows and as long as programming is fun for you. 

I think, having the data stored in an xarray would be more professional, but it is postponable, because the status quo is working. @alexlib: Thanks for offering support, by the way! 

If new features cause a debugging nightmare, this might be an indication for too complicate code. As discussed before, simple, easy and extensible code is a feature of this GUI! If the integration of a new format leads to clumsy code, we could think about writing an extra module for doing nothing but converting any PIV format to any other PIV format (dictionary, csv, NetCDF, LaVision, xarray, whatever). This might be even more useful than integrating it into the GUI, because this module could be reused more easily in another context (in a Jupyter notebook, by any individual script, and of course by the Tkinter-GUI). I still very much like the UNIX philosophy:

- Write programs that do one thing and do it well.
- Write programs to work together.
- Write programs to handle text streams, because that is a universal interface.

The third point is even an argument for keeping the csv format as the main format.

Regards!

Peter

**Comment by alexlib** [comment]
Absolutely +1

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by ErichZimmer** [comment]
Dear Prof. Vennemann,

Did you get the icon images to work with the pip insulations? As far as I know, the last time I installed it, it worked. The reason I am asking is because I am designing a proper icon for the GUI to make it look more professional. Other than that, it's all cleaning up code and making it more readable from there. 

Regards,
Erich Zimmer

**Comment by eguvep** [comment]
Dear Erich,
I returned on that point and found that `include_package_data = True,` in `setup.py` did the job!
You can go ahead designing a nice icon. There is already something to start with: http://www.openpiv.net/images/openpiv_arrow.png although not square.
Take care!
Peter

**Comment by ErichZimmer** [comment]
Dear Prof. Vennemann,

I am still a little worried about the automatic documentation for the GUI parameters. After doing more testing and getting the code to look nice(r), I counted some eleven different widget types. Here is a short list (since I am typing this before work ;));

bool: checkbox inside the main label frame
str[]: listbox
text: text area
labelframe: creates the label frame inside a notebook rider
post_button: Creates the green post process button in postprocessing rider
h-spacer: creates a horizontal spacer inside a label frame
sub_bool: creates a checkbox in a nested label frame
sub_labelframe: creates a nested label frame
sub_h-spacer: creates a horizontal spacer inside of a nester label frame
None: creates rider to place a label frame in
else: creates a text entry

There are also the ints, strings, and other widgets, but I think the automatic documentation doesn't care about those. I noticed that the auto documenter already can select the parameters needed, so I think it shouldn't be too difficult to slightly reconfigure (if needed) the auto documenter so it would exclude the buttons, horizontal spacers, and  label frames.

Regards,
Erich Zimmer

**Comment by eguvep** [comment]
Good morning Erich,
The auto-documentation code is inside the [OpenPivParams.py](https://github.com/OpenPIV/openpiv_tk_gui/blob/master/openpivgui/OpenPivParams.py) file! Scroll down to the very end, there is a function named `generate_parameter_documentation(group)`.
This function is called in the [parameters.rst](https://github.com/OpenPIV/openpiv_tk_gui/blob/master/docs/parameters.rst) file. It is the computron-injection package in Sphinx, that enables Python code in the documentation.
Does that help?
Regards!
Peter

**Comment by ErichZimmer** [comment]
Dear Prof. Vennemann and Prof. Liberzon,

I was thinking about how to "properly" preprocess grey scale images. Right now, preprocessing would be done by normalizing the image to [0, 1] floats, execute preprocessing functions, and finally rescale the images to [0,255] uint. Is this an alright method or is there a better way to perform image preprocessing? I'm just a little bit paranoid that I am making a poor decision from reading so many 'on image preprocessing' articles. 😬

Thanks,
Erich Zimmer

PS, we should try implementing rectangular windowing. I'll be making an issue or pull request soon as I am having quite a bit of trouble on fast flow micro PIV. I looked at the scripting in WinDef.py and process.pyx and it looks like it shouldn't be too difficult, but I may be downplaying the difficulty level due to my relative inexperience with python (became pretty good with Java though 😃).

**Comment by alexlib** [comment]
What we need most are the good tests. Any pre processing algorithm that after that passes the tests should be fine. Later one might look into computational efficiency issues of the transformations between float and integers in 0,255.

**Comment by alexlib** [comment]
Rectangular windows are not very important I think when windef works properly. Maybe I miss something here. 
In any case this should be discussed on another repository - openpiv itself.

**Comment by eguvep** [comment]
Rectangular windows are helpful for archiving high resolution velocity profiles in microchannels and boundary flow profiles.

**Comment by ErichZimmer** [comment]
When performing intensity normalizing to 0,1   it allows for an image in float dtypes to be properly preprocesses without heavy image corruption. When preprocessing is complete, the 0,1 intensities are rescaled to 0,255 so matplotlib would be able to plot the image properly (apparently, it doesn't like floats at all). An additional step on managing max intensities was the addition of intensity capping. All intensities above mean + x * std(), where x is a user defined number which is multiplied by the standard deviation of pixel intensities and added to the mean of the image intensity to get a threshold value, are set to the threshold to remove bright spots. When doing testing with the normalization, preprocessing had no negative effects (that I can notice) and has a positive effect of being 12% faster with the capability of processing images in float data types.

**Comment by alexlib** [comment]
> When performing intensity normalizing to 0,1 it allows for an image in float dtypes to be properly preprocesses without heavy image corruption. When preprocessing is complete, the 0,1 intensities are rescaled to 0,255 so matplotlib would be able to plot the image properly (apparently, it doesn't like floats at all). An additional step on managing max intensities was the addition of intensity capping. All intensities above mean + x * std(), where x is a user defined number which is multiplied by the standard deviation of pixel intensities and added to the mean of the image intensity to get a threshold value, are set to the threshold to remove bright spots. When doing testing with the normalization, preprocessing had no negative effects (that I can notice) and has a positive effect of being 12% faster with the capability of processing images in float data types.

@ErichZimmer - please, we have to split the discussions on different things into different threads or issues. It cannot be all together in one long long discussion on the pull request that we will not merge until we fix the conflicts. 

I suggest to move the discussions on:
* rectangular window
* pre-processing e.g. https://github.com/OpenPIV/openpiv-python/issues/149#issue-729027662
* intensity capping
* everything which is not GUI related 

to the openpiv-python issues and open them one by one, discuss them separately and agree there what is going in and in what order. In parallel, as openpiv-python accepts the new functions (e.g. tools.intensity_capping or preprocessing.intensity_capping), you're very welcome to add here (i.e. at openpiv_tk_gui) the issue for GUI element (button, checkbox, panel, etc.) that allows to include the function in the GUI. 

Otherwise we might have two different versions of openpiv which is a disaster after all. Not everyone will adopt GUI and not everyone will adopt the command line or jupyter notebook way of using it, but all would like to have a clear picture of the preprocessing, processing, post-processing features. 

Thanks, 
Alex

e.g.

**Comment by alexlib** [comment]
https://github.com/OpenPIV/openpiv-python/issues/150

**Comment by eguvep** [comment]
Some other changes have been pushed to the main branch, meanwhile. Please make sure to pull upstream, solve merge conflicts locally, push into this branch and then notify me, so I can merge.
Looking forward to use your changes in the lab!

**Review by eguvep — APPROVED** [review]
Thanks again for improving openpivgui, Erich!

**Comment by ErichZimmer** [comment]
Dear Prof. Vennemann,

I am still working on the GUI, but it is mostly complete right now. I am adding some additional features and an OpenPIV WiDIM algorithm-like output. One feature that is a little troubling is the addition of vector/millisecond output. This feature only works with overlap of 50%. Aside from that, I am just finalizing the progress menu. My computer went wonky today, so I just created a quick backup on GitHub. It should still be function, however, if you want to play around with it.

Regards,
Erich Zimmer

P.S. I got an object masking algorithm to work, but it causes artifacts in the correlation plane due to the masked pixels being set to zero. Eventually, I hope to fill in the masked pixels with artificial data to minimize the aforementioned issue.

**Comment by ErichZimmer** [comment]
Here is how the GUI looks so far for a large size laptop.

General


Preproc


Algor


windowing


Validation1

Note: there was no real need for doing preprocessing as all invalid vectors were successfully interpolated. Additionally, the output was updated to show the user more information about the evaluation.
Postproc

Note, any color can be chosen for the invalid and valid vectors via clicking the color. A pop up menu will then appear. Also, a custom color bar was added. Soon, users can create their own color bar.
Plotting

**Comment by ErichZimmer** [comment]
@eguvep , 
For the main pull request (when I finish all testing and enhancements), should I create a new one or simply merge and close this one?

**Comment by eguvep** [comment]
Simply merge and close this one.

**Comment by eguvep** [comment]
Ups – I think I overlooked your statement in brackets »when I finish all testing and enhancements« and already merged a bit too eager ...

**Comment by eguvep** [comment]
Anyway. I increased the version number to 0.4.0. But I won't tag it as a release, until all testing is finished. So the PyPI version will not be updated until that final changes. You can just start a new pull request for submitting a new version 0.4.1. Please, pull upstream first, because I manually resolved a couple of merge conflicts.

**Comment by eguvep** [comment]
I like, how you implemented the colour choosers. Very elegantly!

**Comment by ErichZimmer** [comment]
@eguvep 
Should we try building the GUI around a native dictionary? This would allow a tremendous amount of flexibility including preprocessing/postprocessing each frames individually or even control how many frames to evaluate. Additionally, the user could be able to customize how the results should be saved and exported. One of my proof-of-concept experiments was based on a database-like system using .npz files and NumPy write/read functions. When loading 10,000 results consisting of x, y, u, v arrays, it took 16 seconds. This, however, could simply be too long. Another negative is that the file filter would only operate for the files loaded into the dictionary and not those of which are in the folder directory containing the images. With the latter issue, I'm not sure if I should proceed and make it a fully functional GUI.

Edit:
I just tried an improved version and got 4.07 seconds for loading 10,000 results ( from 18,250 images), so speed isn't a real big issue here. I am just very worried about the functionality of the file filter.

**Comment by ErichZimmer** [comment]
Yikes, accidentally hit the comment button :(

**Comment by eguvep** [comment]
@ErichZimmer, we should discuss this in a new issue: https://github.com/OpenPIV/openpiv_tk_gui/issues/22
I tried to summarize the discussion, so far, and started a discussion.
